File: 0_1.baekjoon/greedy/java/BOJ8980.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ8980 {
    public static class Delivery {
        int from;
        int to;
        int box;
        Delivery(int from, int to, int box) {
            this.from = from;
            this.to = to;
            this.box = box;
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int N = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());
        
        st = new StringTokenizer(br.readLine());
        int M = Integer.parseInt(st.nextToken());

        Delivery[] works = new Delivery[M];
        for (int i=0;i<M;i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int box = Integer.parseInt(st.nextToken());
            works[i] = new Delivery(from, to, box);
        }

        Arrays.sort(works, (o1, o2) -> o1.to - o2.to);
        int[] delivered = new int[N+1];
        int total = 0;
        for (int i=0;i<M;i++) {
            Delivery info = works[i];
            int max = 0;
            int possible = 0;
            for (int j=info.from;j<info.to;j++) {
                max = Math.max(max, delivered[j]);
            }

            possible = Math.min(C-max, info.box);
            total += possible;
            
            for (int j=info.from;j<info.to;j++) {
                delivered[j] += possible;
            }
        }
        System.out.println(total);
    }
}

File: 0_1.baekjoon/greedy/kotlin/src/BOJ8980.kt
import kotlin.math.max
import kotlin.math.min

fun main() = with(System.`in`.bufferedReader()) {
    val (n, c) = readLine().split(" ").map { it.toInt() }
    val m = readLine().toInt()
    val works =
        List(m) { readLine().split(" ").map { it.toInt() } }
            .sortedWith(
                compareBy(
                    { it[1] },
                    { it[0] })
            )

    val delivery = Array(n + 1) { 0 }
    var answer = 0
    for (i in 0 until m) {
        val work = works[i]
        var max = delivery.slice(work[0]..work[1]).maxOf { d -> d }
        val possible = min(c - max, work[2])
        answer += possible

        for (j in work[0] until work[1]) {
            delivery[j] += possible
        }
    }

    print(answer)
}